### PR TITLE
Updates goal detector info with additions from RoboCup 2021

### PR DESCRIPTION
**Pull request by ysims**
Additions are

- Throwout of goals that are too far away
- Merging of overlapping goal posts

Addresses #156.

### CODE REVIEW

**Review by tayloryoung6396 — COMMENTED** [review]
Also there's a condition to discard clusters above "The cluster intersects the green horizon (we assume goals are higher than the field)" afaik the cluster needs to intersect the green horizon to be a post @Bidski?

**Comment by Bidski** [comment]
> Also there's a condition to discard clusters above "The cluster intersects the green horizon (we assume goals are higher than the field)" afaik the cluster needs to intersect the green horizon to be a post @Bidski? 

Im pretty sure we require that clusters intersect the green horizon. It they dont they either cant be a goal post or they cant belong to our field

File: src/book/02-system/02-subsystems/05-vision.mdx [hunk 1]
```diff
@@ -111,7 +111,9 @@ All goal post edge points (points that are goal points that have at least one ne
...
 
-The bottom centre point of each cluster is then found by averaging the edge points. This is the point where we measure the distance from. If there is more than one cluster, i.e. more than one goal post, an attempt is made to pair up the goal posts. This is done by calculating the distances between the posts and pairing up goal posts which are close together. **Leftness** and **rightness** is assigned to each post in a pair based on their positions.
+From the remaining clusters, clusters are merged if they lie on the same vertical vector. This is to prevent detection of the same post twice and to prevent detection of back posts. The bottom centre point of each cluster is then found by averaging the edge points. This is the point where we measure the distance from. If this distance is too large, as defined in the configuration file, the goal post is discarded.
+
+If there is more than one cluster, i.e. more than one goal post, an attempt is made to pair up the goal posts. This is done by calculating the distances between the posts and pairing up goal posts that are close together. **Leftness** and **rightness** is assigned to each post in a pair based on their positions.
```
[the remark this answers is not in the document]
Comment (Bidski) [reply src/book/02-system/02-subsystems/05-vision.mdx:116]: We dont send pairs. Localisation doesnt use that information. Although localisation probably should use it (and we probably should send pairs when we know it)

**Comment by ysims** [comment]
> > Also there's a condition to discard clusters above "The cluster intersects the green horizon (we assume goals are higher than the field)" afaik the cluster needs to intersect the green horizon to be a post @Bidski?
> 
> Im pretty sure we require that clusters intersect the green horizon. It they dont they either cant be a goal post or they cant belong to our field

"Clusters are discarded if they **fail** to meet any of the following criteria"
I understand if it's easy to miss that word, the way that section is written isn't intuitive. I might change so it's the intuitive way around.

File: src/book/02-system/02-subsystems/05-vision.mdx (hunk 1 above)
[the remark this answers is not in the document]
Comment (ysims) [reply src/book/02-system/02-subsystems/05-vision.mdx:116]: You mean NUbook is wrong!! 😱 
I'll fix that part up in this PR then too

[the remark this answers is not in the document]
Comment (tayloryoung6396) [reply src/book/02-system/02-subsystems/05-vision.mdx:116]: > > > Also there's a condition to discard clusters above "The cluster intersects the green horizon (we assume goals are higher than the field)" afaik the cluster needs to intersect the green horizon to be a post @Bidski?
> > 
> > 
> > Im pretty sure we require that clusters intersect the green horizon. It they dont they either cant be a goal post or they cant belong to our field
> 
> "Clusters are discarded if they **fail** to meet any of the following criteria" I understand if it's easy to miss that word, the way that section is written isn't intuitive. I might change so it's the intuitive way around.

I understand that but **we assume goals are higher than the field** this also doesn't make sense
"Clusters are discarded if they fail to meet a criteria -> fails to meet an intersection ->discarded but you can't intersect and assume goals are higher than the field. I know it doesn't say "only higher" but the sentence feels contradictive

**Comment by tayloryoung6396** [comment]
> > > Also there's a condition to discard clusters above "The cluster intersects the green horizon (we assume goals are higher than the field)" afaik the cluster needs to intersect the green horizon to be a post @Bidski?
> > 
> > 
> > Im pretty sure we require that clusters intersect the green horizon. It they dont they either cant be a goal post or they cant belong to our field
> 
> "Clusters are discarded if they **fail** to meet any of the following criteria" I understand if it's easy to miss that word, the way that section is written isn't intuitive. I might change so it's the intuitive way around.

I understand that but **we assume goals are higher than the field** this also doesn't make sense
"Clusters are discarded if they fail to meet a criteria -> fails to meet an intersection ->discarded but you can't intersect and assume goals are higher than the field. I know it doesn't say "only higher" but the sentence feels contradictive

File: src/book/02-system/02-subsystems/05-vision.mdx (hunk 1 above)
[the remark this answers is not in the document]
Comment (ysims) [reply src/book/02-system/02-subsystems/05-vision.mdx:116]: Actually I guess upon reading it it's just a big vague. I'll clarify it.

[the remark this answers is not in the document]
Comment (tayloryoung6396) [reply src/book/02-system/02-subsystems/05-vision.mdx:116]: NUbook isn't **wrong** but it's misleading.. we do make "pairs" not really exclusive pairs.. but apply left and right labels.. (this is kinda confusing without a picture) but we don't USE pairs. The system doesn't know anything about the pairs other than "this is likely a left post"

**Comment by ysims** [comment]
> > > > Also there's a condition to discard clusters above "The cluster intersects the green horizon (we assume goals are higher than the field)" afaik the cluster needs to intersect the green horizon to be a post @Bidski?
> > > 
> > > 
> > > Im pretty sure we require that clusters intersect the green horizon. It they dont they either cant be a goal post or they cant belong to our field
> > 
> > 
> > "Clusters are discarded if they **fail** to meet any of the following criteria" I understand if it's easy to miss that word, the way that section is written isn't intuitive. I might change so it's the intuitive way around.
> 
> I understand that but **we assume goals are higher than the field** this also doesn't make sense "Clusters are discarded if they fail to meet a criteria -> fails to meet an intersection ->discarded but you can't intersect and assume goals are higher than the field. I know it doesn't say "only higher" but the sentence feels contradictive

OHHHH I get you now. I'll think of a better way to phrase that.

**Comment by tayloryoung6396** [comment]
> > > > > > Also there's a condition to discard clusters above "The cluster intersects the green horizon (we assume goals are higher than the field)" afaik the cluster needs to intersect the green horizon to be a post @Bidski?
> > > > > 
> > > > > 
> > > > > Im pretty sure we require that clusters intersect the green horizon. It they dont they either cant be a goal post or they cant belong to our field
> > > > 
> > > > 
> > > > "Clusters are discarded if they **fail** to meet any of the following criteria" I understand if it's easy to miss that word, the way that section is written isn't intuitive. I might change so it's the intuitive way around.
> > > 
> > > 
> > > I understand that but **we assume goals are higher than the field** this also doesn't make sense "Clusters are discarded if they fail to meet a criteria -> fails to meet an intersection ->discarded but you can't intersect and assume goals are higher than the field. I know it doesn't say "only higher" but the sentence feels contradictive
> > 
> > 
> > OHHHH I get you now. I'll think of a better way to phrase that.
> 
> You should also confirm order of operations in the code. Do we merge clusters before or after we check the horizon intersection?

We check sides and then merge.. seems like a bad order

**Comment by ysims** [comment]
> You should also confirm order of operations in the code. Do we merge clusters before or after we check the horizon intersection?
We do the green horizon check first (`check_green_horizon_side`?), then after it finds the overlapping clusters and merges them. Everything else should be chronological as well unless I've messed up somewhere.

**Comment by ysims** [comment]
Hm I guess do we want to change it around then - overlapping clusters and then green horizon? 
That kind of does make more sense to me. Otherwise if your post detection is split in two, it might get chucked out.

**Comment by Bidski** [comment]
> > You should also confirm order of operations in the code. Do we merge clusters before or after we check the horizon intersection?
> 
> We do the green horizon check first (`check_green_horizon_side`?), then after it finds the overlapping clusters and merges them. Everything else should be chronological as well unless I've messed up somewhere.
> 
> Hm I guess do we want to change it around then - overlapping clusters and then green horizon?
That kind of does make more sense to me. Otherwise if your post detection is split in two, it might get chucked out.

Nah green horizon check should be done first. Any post that is entirely above the green horizon isn't on our field, and any post that is entirely below the green horizon can't actually be a goal post. In both cases we don't want to consider to consider those clusters any further. 

What this leaves is the case where the back post of the goal was being recognised as a third post directly behind one of the front posts, and the pair matching system managed to find two pairs of goal posts from those 3 posts. In this sort of scenario all 3 posts meet our criteria for a goal post (intersects the green horizon and has a minimum number of points) so we have no other means by which to resolve the confusion, other than recognising that you cant "see" one post if it is directly behind another -- hence the merging of the clusters.

**Comment by tayloryoung6396** [comment]
> > > You should also confirm order of operations in the code. Do we merge clusters before or after we check the horizon intersection?
> > 
> > 
> > We do the green horizon check first (`check_green_horizon_side`?), then after it finds the overlapping clusters and merges them. Everything else should be chronological as well unless I've messed up somewhere.
> > Hm I guess do we want to change it around then - overlapping clusters and then green horizon?
> > That kind of does make more sense to me. Otherwise if your post detection is split in two, it might get chucked out.
> 
> Nah green horizon check should be done first. Any post that is entirely above the green horizon isn't on our field, and any post that is entirely below the green horizon can't actually be a goal post. In both cases we don't want to consider to consider those clusters any further.
> 
> What this leaves is the case where the back post of the goal was being recognised as a third post directly behind one of the front posts, and the pair matching system managed to find two pairs of goal posts from those 3 posts. In this sort of scenario all 3 posts meet our criteria for a goal post (intersects the green horizon and has a minimum number of points) so we have no other means by which to resolve the confusion, other than recognising that you cant "see" one post if it is directly behind another -- hence the merging of the clusters.

I mean it also discards correct posts that have clusters detected both above and below GH that don't have a connecting node (a robot arm or something splitting the clusters). But this is probably an edge case where it's safer to do it how we're doing it

**Comment by ysims** [comment]
Ok, in that case there's no need for any changes to the GoalDetector module and this PR should be good now.

File: src/book/02-system/02-subsystems/05-vision.mdx (hunk 1 above)
[the remark this answers is not in the document]
Comment (ysims) [reply src/book/02-system/02-subsystems/05-vision.mdx:116]: Added in a bit below this paragraph about the message not having pair info.

File: src/book/02-system/02-subsystems/05-vision.mdx [hunk 2]
```diff
@@ -52,11 +52,11 @@ After clustering has finished, one cluster will remain which should represent **
...
+The ball detector forms **clusters** out of all the points that the Visual Mesh has identified as ball points. Specifically, the clusters are formed from Visual Mesh points that are identified as being a ball point but have at least $1$ neighbour which is **not** a ball point. This allows us to form clusters of ball edge points. Clusters are discarded if
 
-- The cluster is below the green horizon.
+- The cluster is not below the green horizon.
 
-- The cluster is larger than a given threshold.
+- The cluster is smaller than a given threshold.
 
 Remaining clusters are fitted with a **circular cone**. The **cone axis** is determined from the line segment between the centre of the camera and the centre of the ball, determined by averaging all of the ball edge points. The **radius** of the cone is determined by the maximum distance between the ball edge points and the centre of the ball. Clusters are **discarded** if they fail to meet any of the following criteria
```
Comment (tayloryoung6396) [inline src/book/02-system/02-subsystems/05-vision.mdx:61]: The math in these sections renders pretty bad, there isn't enough vertical distance between the different equations, making the equation hard to read

**Review by tayloryoung6396 — CHANGES_REQUESTED** [review]

File: src/book/02-system/02-subsystems/05-vision.mdx [hunk 3]
```diff
@@ -52,11 +52,11 @@ After clustering has finished, one cluster will remain which should represent **
...
 
-The ball detector forms **clusters** out of all the points that the Visual Mesh has identified as ball points. Specifically, the clusters are formed from Visual Mesh points that are identified as being a ball point but have at least $1$ neighbour which is **not** a ball point. This allows us to form clusters of ball edge points. Clusters are discarded if they fail to meet any of the following criteria
+The ball detector forms **clusters** out of all the points that the Visual Mesh has identified as ball points. Specifically, the clusters are formed from Visual Mesh points that are identified as being a ball point but have at least $1$ neighbour which is **not** a ball point. This allows us to form clusters of ball edge points. Clusters are discarded if
```
[the remark this answers is not in the document]
Comment (ysims) [reply src/book/02-system/02-subsystems/05-vision.mdx]: This PR is getting very scope creepy, but I have gone and added this in for all mentions I could find including goals and field lines (green horizon section).

File: src/book/02-system/02-subsystems/05-vision.mdx (hunk 2 above)
Comment (ysims) [reply src/book/02-system/02-subsystems/05-vision.mdx:61]: Yeeeah they're not great. Not relevant for this particular PR but I did make a PR for this (#201).

**Review by LachlanCourt — APPROVED** [review]

**Comment by ysims** [comment]
@tayloryoung6396

**Pull request merged**